File: engine/mesh2param/profile_fitting.py
"""Bounded spline and regular-polygon hypotheses for prismatic profiles.

The fitters in this module operate only on ordered 2-D mesh evidence.  They use
the same open-uniform, clamped B-spline convention as the CADGraph compiler and
record the residual, complexity penalty, and uncertainty used to select a fit.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any, Literal

import numpy as np
from scipy.interpolate import BSpline
from scipy.optimize import minimize_scalar


@dataclass(frozen=True, slots=True)
class FitUncertainty:
    rms_residual_mm: float
    maximum_residual_mm: float
    sample_count: int
    parameter_iterations: int
    design_condition_number: float

    def to_dict(self) -> dict[str, float | int]:
        return {
            "rmsResidualMm": self.rms_residual_mm,
            "maximumResidualMm": self.maximum_residual_mm,
            "sampleCount": self.sample_count,
            "parameterIterations": self.parameter_iterations,
            "designConditionNumber": self.design_condition_number,
        }


@dataclass(frozen=True, slots=True)
class FitPenalties:
    residual: float
    complexity: float
    endpoint: float
    total: float

    def to_dict(self) -> dict[str, float]:
        return {
            "residual": self.residual,
            "complexity": self.complexity,
            "endpoint": self.endpoint,
            "total": self.total,
        }


@dataclass(frozen=True, slots=True)
class BSplineSegment:
    start: tuple[float, float]
    end: tuple[float, float]
    rms_residual_mm: float
    maximum_residual_mm: float
    length_mm: float
    degree: int
    control_points: tuple[tuple[float, float], ...]
    penalties: FitPenalties
    uncertainty: FitUncertainty
    tangent_to_next_deg: float | None = None

    @property
    def kind(self) -> Literal["bspline"]:
        return "bspline"

    def start_tangent(self) -> np.ndarray:
        vector = np.asarray(self.control_points[1]) - np.asarray(self.control_points[0])
        return np.asarray(vector / np.linalg.norm(vector), dtype=np.float64)

    def end_tangent(self) -> np.ndarray:
        vector = np.asarray(self.control_points[-1]) - np.asarray(self.control_points[-2])
        return np.asarray(vector / np.linalg.norm(vector), dtype=np.float64)


@dataclass(frozen=True, slots=True)
class NominalMeasurement:
    measured_mm: float
    suggested_nominal_mm: float | None
    delta_mm: float | None
    accepted: bool

    @property
    def selected_mm(self) -> float:
        return (
            float(self.suggested_nominal_mm)
            if self.accepted and self.suggested_nominal_mm is not None
            else float(self.measured_mm)
        )

    def to_dict(self) -> dict[str, float | bool | None]:
        return {
            "measuredMm": self.measured_mm,
            "suggestedNominalMm": self.suggested_nominal_mm,
            "deltaMm": self.delta_mm,
            "accepted": self.accepted,
        }


@dataclass(frozen=True, slots=True)
class NominalSnappingPolicy:
    enabled: bool = True
    grid_mm: float = 0.5
    tolerance_mm: float = 0.05

    def snap(self, value_mm: float) -> NominalMeasurement:
        if self.grid_mm <= 0 or self.tolerance_mm < 0:
            raise ValueError("nominal snap grid must be positive and tolerance non-negative")
        if not math.isfinite(value_mm) or value_mm <= 0:
            raise ValueError("nominal measurements must be finite and positive")
        nominal = round(value_mm / self.grid_mm) * self.grid_mm
        delta = float(nominal - value_mm)
        accepted = self.enabled and abs(delta) <= self.tolerance_mm
        return NominalMeasurement(
            float(value_mm),
            float(nominal) if abs(delta) <= self.tolerance_mm else None,
            delta if abs(delta) <= self.tolerance_mm else None,
            accepted,
        )


@dataclass(frozen=True, slots=True)
class RegularPolygonHypothesis:
    side_count: int
    center: tuple[float, float]
    measured_circumdiameter_mm: float
    selected_circumdiameter_mm: float
    phase_angle_rad: float
    clockwise: bool
    maximum_radial_residual_mm: float
    maximum_angular_residual_deg: float
    nominal: NominalMeasurement
    source_points: tuple[tuple[float, float], ...]

    @property
    def points(self) -> tuple[tuple[float, float], ...]:
        direction = -1.0 if self.clockwise else 1.0
        radius = self.selected_circumdiameter_mm * 0.5
        return tuple(
            (
                self.center[0] + radius * math.cos(
                    self.phase_angle_rad + direction * 2.0 * math.pi * index / self.side_count
                ),
                self.center[1] + radius * math.sin(
                    self.phase_angle_rad + direction * 2.0 * math.pi * index / self.side_count
                ),
            )
            for index in range(self.side_count)
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "sideCount": self.side_count,
            "centerMm": list(self.center),
            "measuredCircumdiameterMm": self.measured_circumdiameter_mm,
            "selectedCircumdiameterMm": self.selected_circumdiameter_mm,
            "phaseAngleRad": self.phase_angle_rad,
            "clockwise": self.clockwise,
            "maximumRadialResidualMm": self.maximum_radial_residual_mm,
            "maximumAngularResidualDeg": self.maximum_angular_residual_deg,
            "nominal": self.nominal.to_dict(),
            "sourcePointsMm": [list(point) for point in self.source_points],
        }


def _open_uniform_knots(control_point_count: int, degree: int) -> np.ndarray:
    interior_count = control_point_count - degree - 1
    return np.asarray(
        [0.0] * (degree + 1)
        + [(index + 1) / (interior_count + 1) for index in range(interior_count)]
        + [1.0] * (degree + 1),
        dtype=np.float64,
    )


def _fit_control_points(
    points: np.ndarray,
    parameters: np.ndarray,
    control_point_count: int,
    degree: int,
) -> tuple[np.ndarray, float]:
    basis = BSpline.design_matrix(
        parameters, _open_uniform_knots(control_point_count, degree), degree
    ).toarray()
    controls = np.zeros((control_point_count, 2), dtype=np.float64)
    controls[0] = points[0]
    controls[-1] = points[-1]
    rhs = points - basis[:, [0]] * controls[0] - basis[:, [-1]] * controls[-1]
    interior = basis[:, 1:-1]
    controls[1:-1] = np.linalg.lstsq(interior, rhs, rcond=None)[0]
    return controls, float(np.linalg.cond(interior))


def _fit_one_bspline(
    points: np.ndarray,
    control_point_count: int,
    *,
    degree: int,
    maximum_iterations: int,
) -> BSplineSegment:
    chord_lengths = np.linalg.norm(np.diff(points, axis=0), axis=1)
    cumulative = np.concatenate(([0.0], np.cumsum(chord_lengths)))
    parameters = cumulative / cumulative[-1]
    condition = math.inf
    completed_iterations = 0
    previous_rms = math.inf
    for iteration in range(maximum_iterations):
        controls, condition = _fit_control_points(
            points, parameters, control_point_count, degree
        )
        curve = BSpline(_open_uniform_knots(control_point_count, degree), controls, degree)
        old_parameters = parameters.copy()
        for index in range(1, len(points) - 1):
            lower = old_parameters[index - 1] + 1e-12
            upper = old_parameters[index + 1] - 1e-12
            solution = minimize_scalar(
                lambda value, point=points[index], fitted_curve=curve: float(
                    np.sum((fitted_curve(value) - point) ** 2)
                ),
                bounds=(lower, upper),
                method="bounded",
                options={"xatol": 1e-13, "maxiter": 80},
            )
            parameters[index] = float(solution.x)
        residuals = np.linalg.norm(curve(parameters) - points, axis=1)
        rms = math.sqrt(float(np.mean(residuals**2)))
        completed_iterations = iteration + 1
        if iteration >= 20 and abs(previous_rms - rms) <= 1e-10:
            break
        previous_rms = rms
    controls, condition = _fit_control_points(points, parameters, control_point_count, degree)
    curve = BSpline(_open_uniform_knots(control_point_count, degree), controls, degree)
    residuals = np.linalg.norm(curve(parameters) - points, axis=1)
    rms = math.sqrt(float(np.mean(residuals**2)))
    maximum = float(np.max(residuals))
    dense = curve(np.linspace(0.0, 1.0, 1025))
    length = float(np.sum(np.linalg.norm(np.diff(dense, axis=0), axis=1)))
    residual_penalty = rms + 0.25 * maximum
    complexity_penalty = 0.0025 * (control_point_count - 4)
    endpoint_error = max(
        float(np.linalg.norm(curve(0.0) - points[0])),
        float(np.linalg.norm(curve(1.0) - points[-1])),
    )
    penalties = FitPenalties(
        residual_penalty,
        complexity_penalty,
        endpoint_error,
        residual_penalty + complexity_penalty + endpoint_error,
    )
    uncertainty = FitUncertainty(
        rms,
        maximum,
        len(points),
        completed_iterations,
        condition,
    )
    return BSplineSegment(
        (float(points[0, 0]), float(points[0, 1])),
        (float(points[-1, 0]), float(points[-1, 1])),
        rms,
        maximum,
        length,
        degree,
        tuple((float(point[0]), float(point[1])) for point in controls),
        penalties,
        uncertainty,
    )


def fit_bounded_bspline(
    points: np.ndarray,
    *,
    degree: int = 3,
    minimum_control_points: int = 4,
    maximum_control_points: int = 8,
    rms_tolerance_mm: float = 0.03,
    maximum_tolerance_mm: float = 0.075,
    maximum_iterations: int = 100,
) -> BSplineSegment:
    """Fit a deterministic contract-bounded open-uniform B-spline segment."""

    points = np.asarray(points, dtype=np.float64)
    if points.ndim != 2 or points.shape[1] != 2 or len(points) < 4:
        raise ValueError("B-spline evidence must contain at least four 2-D points")
    if not np.all(np.isfinite(points)):
        raise ValueError("B-spline evidence contains non-finite coordinates")
    if not 1 <= degree <= 3 or minimum_control_points < degree + 1:
        raise ValueError("B-spline degree and control-point bounds are invalid")
    if maximum_control_points > 8 or minimum_control_points > maximum_control_points:
        raise ValueError("B-spline control-point count must remain within 4..8")
    if np.any(np.linalg.norm(np.diff(points, axis=0), axis=1) <= 1e-12):
        raise ValueError("B-spline evidence contains coincident consecutive points")
    fits = [
        _fit_one_bspline(
            points,
            count,
            degree=degree,
            maximum_iterations=maximum_iterations,
        )
        for count in range(minimum_control_points, maximum_control_points + 1)
    ]
    accepted = [
        fit
        for fit in fits
        if fit.rms_residual_mm <= rms_tolerance_mm
        and fit.maximum_residual_mm <= maximum_tolerance_mm
    ]
    if not accepted:
        best = min(fits, key=lambda fit: fit.penalties.total)
        raise ValueError(
            "bounded B-spline fit exceeds residual limits: "
            f"rms={best.rms_residual_mm:g} mm, max={best.maximum_residual_mm:g} mm"
        )
    return min(
        accepted,
        key=lambda fit: (fit.penalties.total, len(fit.control_points), fit.control_points),
    )


def fit_regular_polygon(
    points: np.ndarray,
    *,
    snapping: NominalSnappingPolicy | None = None,
    minimum_sides: int = 3,
    maximum_sides: int = 12,
    radial_tolerance_mm: float = 0.05,
    angular_tolerance_deg: float = 1.0,
) -> RegularPolygonHypothesis:
    """Recognize one ordered regular polygon and apply the explicit snap policy."""

    points = np.asarray(points, dtype=np.float64)
    count = len(points)
    if points.ndim != 2 or points.shape[1] != 2 or not minimum_sides <= count <= maximum_sides:
        raise ValueError("regular-polygon evidence has an unsupported vertex count")
    center = np.mean(points, axis=0)
    radial = points - center
    radii = np.linalg.norm(radial, axis=1)
    measured_radius = float(np.mean(radii))
    maximum_radial = float(np.max(np.abs(radii - measured_radius)))
    signed_area = 0.5 * float(
        np.sum(points[:, 0] * np.roll(points[:, 1], -1) - points[:, 1] * np.roll(points[:, 0], -1))
    )
    clockwise = signed_area < 0
    direction = -1.0 if clockwise else 1.0
    angles = np.unwrap(np.arctan2(radial[:, 1], radial[:, 0]))
    expected = angles[0] + direction * 2.0 * math.pi * np.arange(count) / count
    angular_error = np.angle(np.exp(1j * (angles - expected)))
    maximum_angular = math.degrees(float(np.max(np.abs(angular_error))))
    if maximum_radial > radial_tolerance_mm or maximum_angular > angular_tolerance_deg:
        raise ValueError(
            "loop is not a regular polygon: "
            f"radial={maximum_radial:g} mm, angular={maximum_angular:g} deg"
        )
    nominal = (snapping or NominalSnappingPolicy()).snap(2.0 * measured_radius)
    return RegularPolygonHypothesis(
        count,
        (float(center[0]), float(center[1])),
        2.0 * measured_radius,
        nominal.selected_mm,
        float(angles[0]),
        clockwise,
        maximum_radial,
        maximum_angular,
        nominal,
        tuple((float(point[0]), float(point[1])) for point in points),
    )


__all__ = [
    "BSplineSegment",
    "FitPenalties",
    "FitUncertainty",
    "NominalMeasurement",
    "NominalSnappingPolicy",
    "RegularPolygonHypothesis",
    "fit_bounded_bspline",
    "fit_regular_polygon",
]

File: apps/web/src/local/geometry/profileFitting.ts
import { Matrix, SingularValueDecomposition } from "ml-matrix";

import type { Vec2 } from "./sections";

export interface BrowserBSplinePrimitive {
  kind: "bspline";
  start: Vec2;
  end: Vec2;
  degree: number;
  controlPoints: Vec2[];
  rms: number;
  maximum: number;
  condition: number;
}

export interface BrowserRegularPolygon {
  sideCount: number;
  center: Vec2;
  circumdiameter: number;
  points: Vec2[];
  clockwise: boolean;
  maximumRadialResidual: number;
  maximumAngularResidualDeg: number;
}

function openUniformKnots(controlPointCount: number, degree: number): number[] {
  const interiorCount = controlPointCount - degree - 1;
  return [
    ...Array.from({ length: degree + 1 }, () => 0),
    ...Array.from({ length: interiorCount }, (_value, index) => (index + 1) / (interiorCount + 1)),
    ...Array.from({ length: degree + 1 }, () => 1),
  ];
}

function basisRow(parameterValue: number, controlPointCount: number, degree: number, knots: readonly number[]): number[] {
  const parameter = parameterValue >= 1 ? 1 - Number.EPSILON : Math.max(0, parameterValue);
  let values: number[] = Array.from({ length: controlPointCount }, (_value, index) => (
    knots[index]! <= parameter && parameter < knots[index + 1]! ? 1 : 0
  ));
  for (let order = 1; order <= degree; order += 1) {
    values = values.map((_value, index) => {
      const leftDenominator = knots[index + order]! - knots[index]!;
      const rightDenominator = knots[index + order + 1]! - knots[index + 1]!;
      const left = leftDenominator > 0 ? (parameter - knots[index]!) / leftDenominator * values[index]! : 0;
      const right = rightDenominator > 0 && index + 1 < values.length
        ? (knots[index + order + 1]! - parameter) / rightDenominator * values[index + 1]!
        : 0;
      return left + right;
    });
  }
  if (parameterValue >= 1) {
    values.fill(0);
    values[controlPointCount - 1] = 1;
  }
  return values;
}

function evaluateBSpline(controlPoints: readonly Vec2[], degree: number, parameter: number): Vec2 {
  const basis = basisRow(parameter, controlPoints.length, degree, openUniformKnots(controlPoints.length, degree));
  return controlPoints.reduce<Vec2>((result, point, index) => [
    result[0] + point[0] * basis[index]!,
    result[1] + point[1] * basis[index]!,
  ], [0, 0]);
}

function fitControls(points: readonly Vec2[], parameters: readonly number[], count: number, degree: number): { controls: Vec2[]; condition: number } {
  const knots = openUniformKnots(count, degree);
  const rows = parameters.map((parameter) => basisRow(parameter, count, degree, knots));
  const first = points[0]!; const last = points.at(-1)!;
  const interiorRows = rows.map((row) => row.slice(1, -1));
  const rightHand = rows.map((row, index) => [
    points[index]![0] - row[0]! * first[0] - row.at(-1)! * last[0],
    points[index]![1] - row[0]! * first[1] - row.at(-1)! * last[1],
  ]);
  const decomposition = new SingularValueDecomposition(new Matrix(interiorRows), { autoTranspose: true });
  const solved = decomposition.solve(new Matrix(rightHand));
  const controls: Vec2[] = [first];
  for (let row = 0; row < solved.rows; row += 1) controls.push([solved.get(row, 0), solved.get(row, 1)]);
  controls.push(last);
  return { controls, condition: decomposition.condition };
}

function squaredDistance(left: Vec2, right: Vec2): number {
  return (left[0] - right[0]) ** 2 + (left[1] - right[1]) ** 2;
}

function goldenMinimum(objective: (value: number) => number, lowerValue: number, upperValue: number): number {
  const ratio = (Math.sqrt(5) - 1) / 2;
  let lower = lowerValue; let upper = upperValue;
  let left = upper - ratio * (upper - lower); let right = lower + ratio * (upper - lower);
  let leftValue = objective(left); let rightValue = objective(right);
  for (let iteration = 0; iteration < 80 && upper - lower > 1e-12; iteration += 1) {
    if (leftValue <= rightValue) {
      upper = right; right = left; rightValue = leftValue; left = upper - ratio * (upper - lower); leftValue = objective(left);
    } else {
      lower = left; left = right; leftValue = rightValue; right = lower + ratio * (upper - lower); rightValue = objective(right);
    }
  }
  return (lower + upper) / 2;
}

function fitOne(points: readonly Vec2[], count: number, degree: number): BrowserBSplinePrimitive {
  const chordLengths = points.slice(1).map((point, index) => Math.sqrt(squaredDistance(point, points[index]!)));
  const total = chordLengths.reduce((sum, value) => sum + value, 0);
  if (!(total > 1e-12)) throw new Error("B-spline evidence has zero chord length");
  const parameters = [0];
  for (const chord of chordLengths) parameters.push(parameters.at(-1)! + chord / total);
  let controls: Vec2[] = [];
  let previousRms = Number.POSITIVE_INFINITY;
  for (let iteration = 0; iteration < 50; iteration += 1) {
    controls = fitControls(points, parameters, count, degree).controls;
    const previous = [...parameters];
    for (let index = 1; index < points.length - 1; index += 1) {
      const lower = previous[index - 1]! + 1e-12;
      const upper = previous[index + 1]! - 1e-12;
      parameters[index] = goldenMinimum(
        (parameter) => squaredDistance(evaluateBSpline(controls, degree, parameter), points[index]!),
        lower,
        upper,
      );
    }
    const residuals = points.map((point, index) => Math.sqrt(squaredDistance(evaluateBSpline(controls, degree, parameters[index]!), point)));
    const rms = Math.sqrt(residuals.reduce((sum, value) => sum + value * value, 0) / residuals.length);
    if (iteration >= 15 && Math.abs(previousRms - rms) <= 1e-10) break;
    previousRms = rms;
  }
  const finalFit = fitControls(points, parameters, count, degree);
  controls = finalFit.controls;
  const residuals = points.map((point, index) => Math.sqrt(squaredDistance(evaluateBSpline(controls, degree, parameters[index]!), point)));
  return {
    kind: "bspline",
    start: points[0]!,
    end: points.at(-1)!,
    degree,
    controlPoints: controls,
    rms: Math.sqrt(residuals.reduce((sum, value) => sum + value * value, 0) / residuals.length),
    maximum: Math.max(...residuals),
    condition: finalFit.condition,
  };
}

export function fitBoundedBSpline(
  points: readonly Vec2[],
  rmsTolerance: number,
  maximumTolerance: number,
): BrowserBSplinePrimitive {
  if (points.length < 4 || points.length > 4096) throw new Error("B-spline evidence is outside the bounded point count");
  if (points.some((point) => !point.every(Number.isFinite))) throw new Error("B-spline evidence contains non-finite coordinates");
  const fits = Array.from({ length: 5 }, (_value, index) => fitOne(points, index + 4, 3));
  const accepted = fits.filter((fit) => fit.rms <= rmsTolerance && fit.maximum <= maximumTolerance && Number.isFinite(fit.condition));
  if (accepted.length === 0) {
    const best = fits.sort((left, right) => left.rms + 0.25 * left.maximum - (right.rms + 0.25 * right.maximum))[0]!;
    throw new Error(`Bounded B-spline fit exceeds residual limits: rms=${best.rms}, max=${best.maximum}`);
  }
  return accepted.sort((left, right) => (
    left.rms + 0.25 * left.maximum + 0.0025 * (left.controlPoints.length - 4)
    - (right.rms + 0.25 * right.maximum + 0.0025 * (right.controlPoints.length - 4))
  ))[0]!;
}

function wrapAngle(value: number): number {
  return Math.atan2(Math.sin(value), Math.cos(value));
}

export function fitRegularPolygon(points: readonly Vec2[], tolerance: number): BrowserRegularPolygon {
  if (points.length < 3 || points.length > 12) throw new Error("Regular polygon has an unsupported side count");
  const center: Vec2 = [
    points.reduce((sum, point) => sum + point[0], 0) / points.length,
    points.reduce((sum, point) => sum + point[1], 0) / points.length,
  ];
  const radial = points.map((point) => [point[0] - center[0], point[1] - center[1]] as Vec2);
  const radii = radial.map((point) => Math.hypot(...point));
  const radius = radii.reduce((sum, value) => sum + value, 0) / radii.length;
  const maximumRadialResidual = Math.max(...radii.map((value) => Math.abs(value - radius)));
  const area = points.reduce((sum, point, index) => {
    const next = points[(index + 1) % points.length]!;
    return sum + point[0] * next[1] - point[1] * next[0];
  }, 0) / 2;
  const clockwise = area < 0;
  const direction = clockwise ? -1 : 1;
  const phase = Math.atan2(radial[0]![1], radial[0]![0]);
  const maximumAngularResidualDeg = Math.max(...radial.map((point, index) => (
    Math.abs(wrapAngle(Math.atan2(point[1], point[0]) - (phase + direction * 2 * Math.PI * index / points.length))) * 180 / Math.PI
  )));
  if (maximumRadialResidual > tolerance || maximumAngularResidualDeg > 1) {
    throw new Error(`Loop is not regular: radial=${maximumRadialResidual}, angular=${maximumAngularResidualDeg}`);
  }
  const measuredDiameter = radius * 2;
  const nominal = Math.round(measuredDiameter / 0.5) * 0.5;
  const diameter = Math.abs(nominal - measuredDiameter) <= tolerance ? nominal : measuredDiameter;
  return {
    sideCount: points.length,
    center,
    circumdiameter: diameter,
    clockwise,
    maximumRadialResidual,
    maximumAngularResidualDeg,
    points: Array.from({ length: points.length }, (_value, index) => [
      center[0] + diameter / 2 * Math.cos(phase + direction * 2 * Math.PI * index / points.length),
      center[1] + diameter / 2 * Math.sin(phase + direction * 2 * Math.PI * index / points.length),
    ]),
  };
}
